cpp/src/neighbors/ivf_pq/ivf_pq_build.cuh

```diff
@@ -1754,10 +1797,22 @@ auto build(raft::resources const& handle,
       cluster_centers, index.n_lists(), index.dim());
     cuvs::cluster::kmeans::balanced_params kmeans_params;
     kmeans_params.n_iters = params.kmeans_n_iters;
-    kmeans_params.metric  = static_cast<cuvs::distance::DistanceType>((int)index.metric());
+    if (index.metric() == distance::DistanceType::CosineExpanded) {
+      raft::print_device_vector(
+        "non_normalized_build", trainset.data_handle(), index.dim(), std::cout);
+      raft::linalg::row_normalize(
+        handle, trainset_const_view, trainset.view(), raft::linalg::NormType::L2Norm);
+      raft::print_device_vector("normalized_build", trainset.data_handle(), index.dim(), std::cout);
+      kmeans_params.metric = distance::DistanceType::InnerProduct;
```

Comment:
If you fool kmeans into using inner product instead of cosine metric, it probably produces non-normalized cluster centers. This messes up with the cluster selection stage for the cosine metric and thus likely reduces the recall.
Maybe try to normalize cluster centers after you do the training, or adjust index extend and search code to take this into account?

Reply:
The problem is that then cluster center + residual would certainly have a norm greater than 1 and not close to 1.  Since residuals are trained with L2 as the distance metric.

Reply:
The residuals shouldn't be a problem, but centers need to be normalized (or their norms accounted for) during search, so that correct clusters are selected for the fine search.